pkg/fqdn/dnsproxy/proxy_test.go

```diff
@@ -163,131 +206,325 @@ func (s *DNSProxyTestSuite) TearDownSuite(c *C) {
+	c.Assert(err, Equals, nil, Commentf("Could not update with rules"))
+	allowed, err = s.proxy.CheckAllowed(epID_1, dstPort, dstID_1, query)
+	c.Assert(err, Equals, nil, Commentf("Error when checking allowed"))
+	c.Assert(allowed, Equals, false, Commentf("request was allowed when it should be rejected"))
 }
 
-func (s *DNSProxyTestSuite) TestSetRejectReplyNoValidData(c *C) {
-	s.proxy.SetRejectReply("banana")
-	request := new(dns.Msg)
-	request.SetQuestion("notcilium.io.", dns.TypeA)
-	response, _, err := s.dnsTCPClient.Exchange(request, s.proxy.TCPServer.Listener.Addr().String())
-	c.Assert(err, IsNil, Commentf("DNS request from test client returned error when it should be rejected"))
-	c.Assert(response.Rcode, Not(Equals), 100, Commentf("DNS request from test client has an invalid response code"))
+func (s *DNSProxyTestSuite) TestFullPathDependence(c *C) {
+	// Test that we consider each of endpoint ID, destination SecID (via the
+	// selector in L7DataMap), destination port (set in the redirect itself) and
+	// the DNS name.
+	// The rules approximate:
+	//	+------+--------+---------+----------------+
+	//	| From |   To   | DstPort |    DNSNames    |
+	//	+======+========+=========+================+
+	//	| All  | DstID1 |      53 | *.ubuntu.com   |
+	//	| All  | DstID1 |      53 | aws.amazon.com |
+	//	| All  | DstID2 |      53 | cilium.io      |
+	//	| All  | DstID1 |      54 | example.com    |
+	//	+------+--------+---------+----------------+
+	//
+	// Cases:
+	// +------+-------+--------+------+-------------+----------+-------------------------------------+
+	// | Case | From  |   To   | Port |    Query    | Outcome  |               Reason                |
+	// +------+-------+--------+------+-------------+----------+-------------------------------------+
+	// |    1 | EPID1 | DstID1 |   53 | cilium.io   | Allowed  |                                     |
+	// |    2 | EPID1 | DstID1 |   54 | cilium.io   | Rejected | Port 54 only allows example.com     |
+	// |    3 | EPID1 | DstID2 |   53 | cilium.io   | Rejected | DstID2 is not allowed in any policy |
+	// |    4 | EPID1 | DstID1 |   54 | example.com | Allowed  |                                     |
+	// |    5 | EPID2 | DstID1 |   53 | cilium.io   | Rejected | EPID2 is not allowed in any policy  |
+	// +------+-------+--------+------+-------------+----------+-------------------------------------+
+
+	var (
+		rulesPort53 = []policy.L7DataMap{
+			{
+				// Rule
+				//	| All  | DstID1 |      53 | *.ubuntu.com   |
+				cachedFooSelector: api.L7Rules{
+					DNS: []api.PortRuleDNS{
+						{MatchName: "cilium.io."},
+					},
+				},
+			},
+			{
+				// Rule
+				//	| All  | DstID1 |      53 | aws.amazon.com |
+				cachedFooSelector: api.L7Rules{
+					DNS: []api.PortRuleDNS{
+						{MatchName: "aws.amazon.com."},
+					},
+				},
+			},
+			{
+				// Rule
+				//	| All  | DstID2 |      53 | cilium.io      |
+				cachedFooSelector: api.L7Rules{
```

Comment:
The comment above (and the other comments) use `DstID2` to indicate the L3 peer here, but it looks like it's using the same `cachedFooSelector` as for `DstID1`.

Reply:
Oops, you are correct. I'll rename cacheFooSelector to Dst1Selector or something to make it clearer. I'll also add the Dst2 equivalents so this rule works (and fix the test cases)